fiftyone/utils/csv.py

```diff
@@ -6,13 +6,230 @@
+            samples are imported
+        seed (None): a random seed to use when shuffling
+        max_samples (None): a maximum number of samples to import. By default,
+            all samples are imported
+    """
+
+    def __init__(
+        self,
+        dataset_dir=None,
+        data_path=None,
+        labels_path=None,
+        media_field="filepath",
+        fields=None,
+        skip_missing_media=False,
+        include_all_data=False,
+        shuffle=False,
+        seed=None,
+        max_samples=None,
+    ):
+        if dataset_dir is None and data_path is None and labels_path is None:
+            raise ValueError(
+                "At least one of `dataset_dir`, `data_path`, and "
+                "`labels_path` must be provided"
+            )
+
+        data_path = self._parse_data_path(
+            dataset_dir=dataset_dir,
+            data_path=data_path,
+            default="data/",
+        )
+
+        labels_path = self._parse_labels_path(
+            dataset_dir=dataset_dir,
+            labels_path=labels_path,
+            default="labels.csv",
+        )
+
+        super().__init__(
+            dataset_dir=dataset_dir,
+            shuffle=shuffle,
+            seed=seed,
+            max_samples=max_samples,
+        )
+
+        self.data_path = data_path
+        self.labels_path = labels_path
+        self.media_field = media_field
+        self.fields = fields
+        self.skip_missing_media = skip_missing_media
+        self.include_all_data = include_all_data
+
+        self._media_paths_map = None
+        self._rows_map = None
+        self._fields = None
+        self._filepaths = None
+        self._iter_filepaths = None
+        self._num_samples = None
+
+    def __iter__(self):
+        self._iter_filepaths = iter(self._filepaths)
+        return self
+
+    def __len__(self):
+        return self._num_samples
+
+    def __next__(self):
+        filepath = next(self._iter_filepaths)
+
+        row = self._rows_map.get(filepath, None)
+        if row is None:
+            return fos.Sample(filepath=filepath)
+
+        kwargs = {}
+        if isinstance(self._fields, dict):
+            for key, parser in self._fields.items():
+                value = row.get(key, "")
+                if value:
+                    kwargs[key] = parser(value)
+        elif self._fields is not None:
+            for key in self._fields:
+                value = row.get(key, "")
+                if value:
+                    kwargs[key] = value
+        else:
+            for key, value in row.items():
+                if value:
+                    kwargs[key] = value
+
+        return fos.Sample(filepath=filepath, **kwargs)
+
+    @property
+    def has_sample_field_schema(self):
+        return False
+
+    @property
+    def has_dataset_info(self):
+        return False
+
+    def setup(self):
+        media_paths_map = self._load_data_map(self.data_path, recursive=True)
```

Comment:
Our builtin importers are generally implemented using this `_load_data_map()` function. I'll explain offline when we have some free time.